src/transformStateWithClones.js

```diff
@@ -7,7 +7,24 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const newState = { ...state };
+  const arrayStates = [];
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      Object.assign(newState, action.extraData)
+    } else if (action.type === 'keysToRemove') {
+      for (const key of action.keysToRemove) {
+        delete newState[key];
+      }
+    } else {
+      newState = {};
+    }
+
+    arrayStates.push({ ...newState });
```

Comment:
It's better to push a copy of the object to the 'arrayStates' at the end of the loop cycle, but outside of the switch block, after the current action is processed. This way, you avoid repeating the same action in all cases.